### PR TITLE
2261 split login screen

### PR DESCRIPTION
**Pull request by sirpy**
# Description

- add /userExists end point
- move auth limit on /user to storageAPI

### CODE REVIEW

File: src/server/storage/storageAPI.js [hunk 1]
```diff
@@ -18,6 +18,22 @@ const adminAuthenticate = (req, res, next) => {
 }
 
 const setup = (app: Router, gunPublic: StorageAPI, storage: StorageAPI) => {
+  app.use(
+    ['/user/*'],
+    passport.authenticate('jwt', { session: false }),
+    wrapAsync(async (req, res, next) => {
+      const { user, body, log } = req
+      const identifier = get(body, 'user.identifier') || user.loggedInAs
```
Comment (serdiukov-o-nordwhale) [inline src/server/storage/storageAPI.js:26]:
```suggestion
      const { loggedInAs } = user
      const identifier = get(body, 'user.identifier', loggedInAs)
```

File: src/server/storage/storageAPI.js [hunk 2]
```diff
@@ -18,6 +18,22 @@ const adminAuthenticate = (req, res, next) => {
...
+
+      log.debug(`${req.baseUrl} auth:`, { user, body })
+
+      if (user.loggedInAs !== identifier) {
+        log.warn(`Trying to update other user data! ${user.loggedInAs}!==${identifier}`)
+        throw new Error(`Trying to update other user data! ${user.loggedInAs}!==${identifier}`)
```
Comment (serdiukov-o-nordwhale) [inline src/server/storage/storageAPI.js:32]:
```suggestion
      if (loggedInAs !== identifier) {
        log.warn(`Trying to update other user data! ${loggedInAs}!==${identifier}`)
        throw new Error(`Trying to update other user data! ${loggedInAs}!==${identifier}`)
```

File: src/server/storage/storageAPI.js [hunk 3]
```diff
@@ -293,14 +306,52 @@ const setup = (app: Router, gunPublic: StorageAPI, storage: StorageAPI) => {
...
+   */
+  app.post(
+    '/userExists',
+    wrapAsync(async (req, res, next) => {
+      const { identifier, email, mobile } = req.body
+      const existing = await storage.model
+        .find({
+          createdDate: { $exists: true },
+          $or: [{ identifier }, { email: email && sha3(email) }, { mobile: mobile && sha3(mobile) }]
+        })
```
Comment (serdiukov-o-nordwhale) [inline src/server/storage/storageAPI.js:336]:
```suggestion
        })
        .sort('isVerified createdDate')
```

File: src/server/storage/storageAPI.js [hunk 4]
```diff
@@ -293,14 +306,52 @@ const setup = (app: Router, gunPublic: StorageAPI, storage: StorageAPI) => {
...
+        .lean()
+
+      if (existing.length) {
+        //prefer oldest verified account
+        const sorted = sortBy(existing, ['isVerified', 'createdDate'])
```
Comment (serdiukov-o-nordwhale) [inline src/server/storage/storageAPI.js:341]:
```suggestion
        const [{ torusProvider, fullName }] = existing
```

File: src/server/storage/storageAPI.js [hunk 5]
```diff
@@ -293,14 +306,52 @@ const setup = (app: Router, gunPublic: StorageAPI, storage: StorageAPI) => {
...
+        return res.json({
+          ok: 1,
+          found: existing.length,
+          exists: true,
+          provider: sorted[0].torusProvider,
+          fullName: sorted[0].fullName
```
Comment (serdiukov-o-nordwhale) [inline src/server/storage/storageAPI.js:347]:
```suggestion
          provider: torusProvider,
          fullName
```

File: src/server/storage/storageAPI.js [hunk 6]
```diff
@@ -30,7 +46,6 @@ const setup = (app: Router, gunPublic: StorageAPI, storage: StorageAPI) => {
    */
   app.post(
     '/user/add',
-    passport.authenticate('jwt', { session: false }),
```
Comment (serdiukov-o-nordwhale) [inline src/server/storage/storageAPI.js:33]: so we're not using login with temporary creds during signup anymore ?

**Review by serdiukov-o-nordwhale — CHANGES_REQUESTED** [review]

File: src/server/storage/storageAPI.js (hunk 6 above)
Comment (sirpy) [reply src/server/storage/storageAPI.js:33]: we all /user endpoints have automatically authentication. see '/user/*'

File: src/server/storage/storageAPI.js (hunk 4 above)
Comment (sirpy) [reply src/server/storage/storageAPI.js:341]: sort still returns a list and not a single item.
i'm also not sure the database sort is in same order, so i'm just keeping the current working and tested setup

File: src/server/storage/storageAPI.js [hunk 7]
```diff
@@ -1,7 +1,7 @@
 // @flow
 import { Router } from 'express'
 import passport from 'passport'
-import { defaults, omitBy } from 'lodash'
+import { defaults, get, omitBy, sortBy } from 'lodash'
```
Comment (serdiukov-o-nordwhale) [inline src/server/storage/storageAPI.js:4]:
```suggestion
import { defaults, get, omitBy, sortBy, first } from 'lodash'
```

File: src/server/storage/storageAPI.js [hunk 8]
```diff
@@ -293,14 +307,52 @@ const setup = (app: Router, gunPublic: StorageAPI, storage: StorageAPI) => {
...
+      const existing = await storage.model
+        .find({
+          createdDate: { $exists: true },
+          $or: [{ identifier }, { email: email && sha3(email) }, { mobile: mobile && sha3(mobile) }]
+        })
+        .lean()
+
+      if (existing.length) {
+        //prefer oldest verified account
+        const sorted = sortBy(existing, ['isVerified', 'createdDate'])
```
Comment (serdiukov-o-nordwhale) [inline src/server/storage/storageAPI.js:342]:
```suggestion
        const { torusProvider, fullName } = first(sortBy(existing, ['isVerified', 'createdDate']))
```

**Review by serdiukov-o-nordwhale — APPROVED** [review]

**Pull request merged**